README.md

```diff
@@ -98,6 +98,152 @@ allows us to do:
+
+    type Close = Eps;
+
+Any channel whose type is `Chan<E, Eps>` implements the `close()` function that
+closes the connection.
+
+### `Send` and `Recv`
+
+These are the most basic of actions: Transmitting and receiving values. The
+opposite of a `Send` with some type `T` is a `Recv` of type `T`.
+
+    type S = Send<String, Eps>;
+    type R = Recv<String, Eps>; // <S as HasDual>::Dual
+
+A channel of type `Chan<E, Send<T, P>>` implements the function `send(T) → Chan<E, P>`.
+
+A channel of type `Chan<E, Recv<T, Q>>` implements the function `recv() → (T, Chan<E, Q>)`.
+
+### `Choose` and `Offer`
+
+There is the option of making choices in the protocol, for one process to
+inform the other of a decision. The `Choose` and `Offer` constructs model such
+choices.
+
+    type C = Choose<Send<String, Eps>, Recv<String, Eps>>;
+    type O = Offer<Recv<String, Eps>, Send<String, Eps>; // <C as Hasdual>::Dual
+
+A channel of type `Chan<E, Choose<P, Q>>` implements _two_ functions:
+
+ * `sel1() → Chan<E, P>`
+ * `sel2() → Chan<E, Q>`
+
+that communicates the choice of protocols `P` and `Q` to the other process.
+
+A channel of type `Chan<E, Offer<P, Q>>` implements `offer() → Branch<Chan<E,
+P>, Chan<E, Q>>`. `Branch` is an enum defined as:
+
+    enum Branch<L, R> {
+        Left(L),
+        Right(R),
+    }
+
+A call to `offer()` should then be matched on to figure out which path the
+other process decided to take.
+
+    match c.offer() {
+        Branch::Left(c) => …,
+        Branch::Right(c) => …,
+    }
+
+### `Rec`, `Var`, `S` and `Z`
+
+The type `Rec` implements the ability the recurse in the protocol, ie provides
+an iterator component. The type `Rec<P>` allows repeating the protocol `P`.
+
+A channel of type `Chan<E, Rec<P>>` implements the function `enter() → Chan<(P,
+E), P>`. What `enter()` does is "store" the protocol `P` in the channel
+environment.
+
+The `Var` construct is then used to reference protocols stored in the
+environment. As the environment is essentially a stack, `Var` takes a counter
+that is modeled as a Peano number. `Var<Z>` points to the top of the stack.
+
+A channel of type `Chan<(P, E), Var<Z>>` implements the function `zero() →
+Chan<(P, E), P>`, ie `Var<Z>` is replaced by `P` at the top of the stack.
+
+    type RS = Rec<Send<String, Var<Z>>>;
+    type RR = Rec<Recv<String, Var<Z>>>; // <RS as HasDual>::Dual
+
+The following program indefinitely sends some string:
+
+    let c: Chan<(), Rec<Send<String, Var<Z>>> = …;
+    let c = c.enter();
+    loop {
+        c = c.send("Hello!".to_string()).zero();
+    }
+
+Protocols in the environment can also be popped from the stack with
+`Var<S<N>>`. This requires there to be at least one protocol in the stack.
+
+A channel of type `Chan<(P, E), Var<S<N>>>` implements the function `succ() ->
+Chan<E, Var<N>>` that peels away the `P` from the environment and `S` in the
+counter.
+
+With all the above, the following protocol should make sense:
```

Comment:
```suggestion
Using the constructions from above, we can construct the following protocol:
```

I think `should make sense` is a little demanding of the reader :-)

Reply:
Changed.